collector/ktls_linux.go

```diff
@@ -0,0 +1,130 @@
+// Copyright 2024 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//go:build !noktls
+// +build !noktls
+
+package collector
+
+import (
+	"fmt"
+
+	"github.com/go-kit/log"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/procfs"
+)
+
+type ktlsCollector struct {
+	fs     procfs.FS
+	logger log.Logger
+}
+
+func init() {
+	registerCollector("ktls", defaultDisabled, NewKTLSCollector)
+}
+
+// NewKTLSCollector returns a new Collector exposing kTLS stats.
+func NewKTLSCollector(logger log.Logger) (Collector, error) {
+	fs, err := procfs.NewFS(*procPath)
+	if err != nil {
+		return nil, fmt.Errorf("failed to open procfs: %w", err)
+	}
+
+	return &ktlsCollector{
+		fs:     fs,
+		logger: logger,
+	}, nil
+}
+
+func (c *ktlsCollector) Update(ch chan<- prometheus.Metric) error {
+	stat, err := c.fs.NewTLSStat()
+	if err != nil {
+		return err
+	}
+
+	ktlsCurrTxSwDesc := prometheus.NewDesc(
+		prometheus.BuildFQName(namespace, "ktls", "tls_curr_tx_sw"),
+		"number of TX sessions currently installed where host handles cryptography",
+		nil, nil,
+	)
+	ktlsCurrRxSwDesc := prometheus.NewDesc(
+		prometheus.BuildFQName(namespace, "ktls", "tls_curr_rx_sw"),
```

Comment:
These metric names are a bit confusing compared to the help text.

Maybe something like this? We try and make metric names more human friendly.
```suggestion
		prometheus.BuildFQName(namespace, "ktls", "receive_sessions"),
```